File: framework/wcf/Extensibility/Transactions/TransactionMessagePropertyUdpTransport/VB/UdpTransport/SampleProfileUdpBinding.vb
'  Copyright (c) Microsoft Corporation. All rights reserved.

Imports System.Configuration
Imports System.ServiceModel
Imports System.ServiceModel.Channels

Namespace Microsoft.ServiceModel.Samples
	''' <summary>
	''' Binding for Udp. This is our "sample profile" for Udp, which uses Text+Soap 1.2 
	''' and allows for variation in Reliability capabilities. If ReliableSessionEnabled is set
	''' then we will layer RM+CompositeDuplex on top of Udp. Otherwise we will just
	''' have Udp on our stack.  
	''' </summary>
	Public Class SampleProfileUdpBinding
		Inherits Binding

        Private reliableSessionEnabled_local As Boolean

        ' private BindingElements
        Private compositeDuplex As CompositeDuplexBindingElement
        Private session As ReliableSessionBindingElement
        Private transport As UdpTransportBindingElement
        Private encoding As MessageEncodingBindingElement

        Public Sub New()
            Initialize()
        End Sub

        Public Sub New(ByVal reliableSessionEnabled As Boolean)
            Me.New()
            Me.ReliableSessionEnabled = reliableSessionEnabled
        End Sub

        Public Sub New(ByVal configurationName As String)
            Me.New()
            ApplyConfiguration(configurationName)
        End Sub

        Public Property OrderedSession() As Boolean
            Get
                Return session.Ordered
            End Get
            Set(ByVal value As Boolean)
                session.Ordered = value
            End Set
        End Property

        Public Property ReliableSessionEnabled() As Boolean
            Get
                Return reliableSessionEnabled_local
            End Get
            Set(ByVal value As Boolean)
                reliableSessionEnabled_local = value
            End Set
        End Property

		Public Overrides ReadOnly Property Scheme() As String
			Get
				Return "soap.udp"
			End Get
		End Property

		Public Property SessionInactivityTimeout() As TimeSpan
			Get
				Return Me.session.InactivityTimeout
			End Get
			Set(ByVal value As TimeSpan)
				Me.session.InactivityTimeout = value
			End Set
		End Property

		Public ReadOnly Property SoapVersion() As EnvelopeVersion
			Get
				Return EnvelopeVersion.Soap12
			End Get
		End Property

		Public Property ClientBaseAddress() As Uri
			Get
				Return Me.compositeDuplex.ClientBaseAddress
			End Get
			Set(ByVal value As Uri)
				Me.compositeDuplex.ClientBaseAddress = value
			End Set
		End Property

		''' <summary>
		''' Create the set of binding elements that make up this binding. 
		''' NOTE: order of binding elements is important.
		''' </summary>
		''' <returns></returns>
		Public Overrides Function CreateBindingElements() As BindingElementCollection
			Dim bindingElements As New BindingElementCollection()

			If ReliableSessionEnabled Then
				bindingElements.Add(session)
				bindingElements.Add(compositeDuplex)
			End If

			bindingElements.Add(encoding)
			bindingElements.Add(transport)

			Return bindingElements.Clone()
		End Function

		Private Sub ApplyConfiguration(ByVal configurationName As String)
            Dim section As SampleProfileUdpBindingCollectionElement = CType(ConfigurationManager.GetSection(UdpConstants.UdpBindingSectionName), 
                                                                            SampleProfileUdpBindingCollectionElement)
			Dim element As SampleProfileUdpBindingConfigurationElement = section.Bindings(configurationName)
			If element Is Nothing Then
                Throw New ConfigurationErrorsException(String.Format(System.Globalization.CultureInfo.CurrentCulture,
                                                                     "There is no binding named {0} at {1}.",
                                                                     configurationName,
                                                                     section.BindingName))
			Else
				element.ApplyConfiguration(Me)
			End If
		End Sub

		Private Sub Initialize()
			transport = New UdpTransportBindingElement()
			session = New ReliableSessionBindingElement()
			compositeDuplex = New CompositeDuplexBindingElement()
			encoding = New TextMessageEncodingBindingElement()
		End Sub

		'initialize a SampleProfileUdpBinding from the info collected in a ReliableSessionBindingElement if one is present.
        Private Sub InitializeFrom(ByVal udpTransportBindingElement As UdpTransportBindingElement,
                                   ByVal textMessageEncodingBindingElement As TextMessageEncodingBindingElement,
                                   ByVal reliableSessionBindingElement As ReliableSessionBindingElement,
                                   ByVal compositeDuplexBindingElement As CompositeDuplexBindingElement)
            Me.transport.Multicast = udpTransportBindingElement.Multicast
            Me.transport.MaxBufferPoolSize = udpTransportBindingElement.MaxBufferPoolSize
            Me.transport.MaxReceivedMessageSize = udpTransportBindingElement.MaxReceivedMessageSize

            CType(Me.encoding, TextMessageEncodingBindingElement).WriteEncoding = textMessageEncodingBindingElement.WriteEncoding
            textMessageEncodingBindingElement.ReaderQuotas.CopyTo((CType(Me.encoding, TextMessageEncodingBindingElement)).ReaderQuotas)

            Me.ReliableSessionEnabled = reliableSessionBindingElement IsNot Nothing

            If reliableSessionBindingElement IsNot Nothing Then
                Me.SessionInactivityTimeout = reliableSessionBindingElement.InactivityTimeout
                Me.OrderedSession = reliableSessionBindingElement.Ordered
            End If

            If compositeDuplexBindingElement IsNot Nothing Then
                Me.ClientBaseAddress = compositeDuplexBindingElement.ClientBaseAddress
            End If
        End Sub

		'try to create a SampleProfileUdpBinding from the collection of BindingElement
		'returns true if it is possible, with the resulting binding.
        Public Shared Function TryCreate(ByVal elements As BindingElementCollection,
                                         <System.Runtime.InteropServices.Out()> ByRef binding As Binding) As Boolean
            binding = Nothing
            If elements.Count > 4 Then
                Return False
            End If

            Dim reliableSessionBindingElement As ReliableSessionBindingElement = Nothing
            Dim compositeDuplexBindingElement As CompositeDuplexBindingElement = Nothing
            Dim textMessageEncodingBindingElement As TextMessageEncodingBindingElement = Nothing
            Dim udpTransportBindingElement As UdpTransportBindingElement = Nothing

            For Each element As BindingElement In elements
                If TypeOf element Is CompositeDuplexBindingElement Then
                    compositeDuplexBindingElement = TryCast(element, CompositeDuplexBindingElement)
                ElseIf TypeOf element Is TransportBindingElement Then
                    udpTransportBindingElement = TryCast(element, UdpTransportBindingElement)
                ElseIf TypeOf element Is TextMessageEncodingBindingElement Then
                    textMessageEncodingBindingElement = TryCast(element, TextMessageEncodingBindingElement)
                ElseIf TypeOf element Is ReliableSessionBindingElement Then
                    reliableSessionBindingElement = TryCast(element, ReliableSessionBindingElement)
                Else
                    Return False
                End If
            Next element

            If udpTransportBindingElement Is Nothing Then
                Return False
            End If
            If textMessageEncodingBindingElement Is Nothing Then
                Return False
            End If

            If ((reliableSessionBindingElement IsNot Nothing) AndAlso
                (compositeDuplexBindingElement Is Nothing)) OrElse
            ((reliableSessionBindingElement Is Nothing) AndAlso
             (compositeDuplexBindingElement IsNot Nothing)) Then
                Return False
            End If

            Dim sampleProfileUdpBinding As New SampleProfileUdpBinding()
            sampleProfileUdpBinding.InitializeFrom(udpTransportBindingElement, textMessageEncodingBindingElement,
                                                   reliableSessionBindingElement, compositeDuplexBindingElement)
            If Not sampleProfileUdpBinding.IsBindingElementsMatch(udpTransportBindingElement,
                                                                  textMessageEncodingBindingElement,
                                                                  reliableSessionBindingElement,
                                                                  compositeDuplexBindingElement) Then
                Return False
            End If

            binding = sampleProfileUdpBinding
            Return True
        End Function

        Private Function IsBindingElementsMatch(ByVal udpTransportBindingElement As UdpTransportBindingElement,
                                                ByVal textMessageEncodingBindingElement As TextMessageEncodingBindingElement,
                                                ByVal reliableSessionBindingElement As ReliableSessionBindingElement,
                                                ByVal compositeDuplexBindingElement As CompositeDuplexBindingElement) As Boolean
            If Not IsTransportMatch(Me.transport, udpTransportBindingElement) Then
                Return False
            End If

            If Not IsEncodingMatch(Me.encoding, textMessageEncodingBindingElement) Then
                Return False
            End If

            If Me.ReliableSessionEnabled Then
                If Not IsSessionMatch(Me.session, reliableSessionBindingElement) Then
                    Return False
                End If
                If compositeDuplexBindingElement IsNot Nothing Then
                    If Not IsCompositeDuplexMatch(Me.compositeDuplex, compositeDuplexBindingElement) Then
                        Return False
                    End If
                Else
                    Return False
                End If
            ElseIf reliableSessionBindingElement IsNot Nothing Then
                Return False
            End If

            Return True
        End Function

		Private Function IsTransportMatch(ByVal a As BindingElement, ByVal b As BindingElement) As Boolean
			If b Is Nothing Then
				Return False
			End If

			Dim transportA As UdpTransportBindingElement = TryCast(a, UdpTransportBindingElement)
			Dim transportB As UdpTransportBindingElement = TryCast(b, UdpTransportBindingElement)

			If transportB Is Nothing Then
				Return False
			End If
			If transportA.MaxBufferPoolSize <> transportB.MaxBufferPoolSize Then
				Return False
			End If
			If transportA.MaxReceivedMessageSize <> transportB.MaxReceivedMessageSize Then
				Return False
			End If
			If transportA.Multicast <> transportB.Multicast Then
				Return False
			End If

			Return True
		End Function

		Private Function IsEncodingMatch(ByVal a As BindingElement, ByVal b As BindingElement) As Boolean
			If b Is Nothing Then
				Return False
			End If

			Dim messageEncodingBindingElement As MessageEncodingBindingElement = TryCast(b, MessageEncodingBindingElement)
			If messageEncodingBindingElement Is Nothing Then
				Return False
			End If

			Dim textA As TextMessageEncodingBindingElement = TryCast(a, TextMessageEncodingBindingElement)
			Dim textB As TextMessageEncodingBindingElement = TryCast(b, TextMessageEncodingBindingElement)
			If textB Is Nothing Then
				Return False
			End If
			If textA.MaxReadPoolSize <> textB.MaxReadPoolSize Then
				Return False
			End If
			If textA.MaxWritePoolSize <> textB.MaxWritePoolSize Then
				Return False
			End If

			' compare XmlDictionaryReaderQuotas
			If textA.ReaderQuotas.MaxStringContentLength <> textB.ReaderQuotas.MaxStringContentLength Then
				Return False
			End If
			If textA.ReaderQuotas.MaxArrayLength <> textB.ReaderQuotas.MaxArrayLength Then
				Return False
			End If
			If textA.ReaderQuotas.MaxBytesPerRead <> textB.ReaderQuotas.MaxBytesPerRead Then
				Return False
			End If
			If textA.ReaderQuotas.MaxDepth <> textB.ReaderQuotas.MaxDepth Then
				Return False
			End If
			If textA.ReaderQuotas.MaxNameTableCharCount <> textB.ReaderQuotas.MaxNameTableCharCount Then
				Return False
			End If

			If textA.WriteEncoding.EncodingName <> textB.WriteEncoding.EncodingName Then
				Return False
			End If
			If Not IsMessageVersionMatch(textA.MessageVersion, textB.MessageVersion) Then
				Return False
			End If

			Return True
		End Function

		Private Function IsMessageVersionMatch(ByVal a As MessageVersion, ByVal b As MessageVersion) As Boolean
			If b Is Nothing Then
				Throw New ArgumentNullException("b")
			End If
			If a.Addressing Is Nothing Then
				Throw New InvalidOperationException("MessageVersion.Addressing cannot be null")
			End If

            If a.Envelope IsNot b.Envelope Then
                Return False
            End If
			'if (a.Addressing.Namespace != b.Addressing.Namespace)
			'{
			'    return false;
			'}

			Return True
		End Function

		Private Function IsSessionMatch(ByVal a As BindingElement, ByVal b As BindingElement) As Boolean
			If b Is Nothing Then
				Return False
			End If

			Dim sessionA As ReliableSessionBindingElement = TryCast(a, ReliableSessionBindingElement)
			Dim sessionB As ReliableSessionBindingElement = TryCast(b, ReliableSessionBindingElement)

			If sessionB Is Nothing Then
				Return False
			End If
			If sessionA.AcknowledgementInterval <> sessionB.AcknowledgementInterval Then
				Return False
			End If
			If sessionA.FlowControlEnabled <> sessionB.FlowControlEnabled Then
				Return False
			End If
			If sessionA.InactivityTimeout <> sessionB.InactivityTimeout Then
				Return False
			End If
			If sessionA.MaxPendingChannels <> sessionB.MaxPendingChannels Then
				Return False
			End If
			If sessionA.MaxRetryCount <> sessionB.MaxRetryCount Then
				Return False
			End If
			If sessionA.MaxTransferWindowSize <> sessionB.MaxTransferWindowSize Then
				Return False
			End If
			If sessionA.Ordered <> sessionB.Ordered Then
				Return False
			End If

			Return True
		End Function

		Private Function IsCompositeDuplexMatch(ByVal a As BindingElement, ByVal b As BindingElement) As Boolean
			If b Is Nothing Then
				Return False
			End If

			Dim duplexA As CompositeDuplexBindingElement = TryCast(a, CompositeDuplexBindingElement)
			Dim duplexB As CompositeDuplexBindingElement = TryCast(b, CompositeDuplexBindingElement)
			If duplexB Is Nothing Then
				Return False
			End If

			Return (duplexB.ClientBaseAddress Is duplexA.ClientBaseAddress)
		End Function
	End Class
End Namespace

File: framework/wcf/Extensibility/Transport/Udp/CS/UdpTransport/SampleProfileUdpBinding.cs
//----------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
//----------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Configuration;
using System.ServiceModel;
using System.ServiceModel.Channels;

namespace Microsoft.Samples.Udp
{

    /// <summary>
    /// Binding for Udp. This is our "sample profile" for Udp, which uses Text+Soap 1.2 
    /// and allows for variation in Reliability capabilities. If ReliableSessionEnabled is set
    /// then we will layer RM+CompositeDuplex on top of Udp. Otherwise we will just
    /// have Udp on our stack.  
    /// </summary>
    public class SampleProfileUdpBinding : Binding
    {
        bool reliableSessionEnabled;

        // private BindingElements
        CompositeDuplexBindingElement compositeDuplex;
        ReliableSessionBindingElement session;
        UdpTransportBindingElement transport;
        MessageEncodingBindingElement encoding;

        public SampleProfileUdpBinding()
        {
            Initialize();
        }

        public SampleProfileUdpBinding(bool reliableSessionEnabled) : this()
        {
            this.ReliableSessionEnabled = reliableSessionEnabled;
        }

        public SampleProfileUdpBinding(string configurationName) : this()
        {
            ApplyConfiguration(configurationName);
        }

        public bool OrderedSession
        {
            get { return session.Ordered; }
            set { session.Ordered = value; }
        }

        public bool ReliableSessionEnabled
        {
            get { return reliableSessionEnabled; }
            set { reliableSessionEnabled = value; }
        }

        public override string Scheme { get { return "soap.udp"; } }

        public TimeSpan SessionInactivityTimeout
        {
            get { return this.session.InactivityTimeout; }
            set { this.session.InactivityTimeout = value; }
        }

        public EnvelopeVersion SoapVersion
        {
            get { return EnvelopeVersion.Soap12; }
        }

        public Uri ClientBaseAddress
        {
            get { return this.compositeDuplex.ClientBaseAddress; }
            set { this.compositeDuplex.ClientBaseAddress = value; }
        }

        /// <summary>
        /// Create the set of binding elements that make up this binding. 
        /// NOTE: order of binding elements is important.
        /// </summary>
        /// <returns></returns>
        public override BindingElementCollection CreateBindingElements()
        {   
            BindingElementCollection bindingElements = new BindingElementCollection();

            if (ReliableSessionEnabled)
            {
                bindingElements.Add(session);
                bindingElements.Add(compositeDuplex);
            }

            bindingElements.Add(encoding);
            bindingElements.Add(transport);

            return bindingElements.Clone();
        }

        void ApplyConfiguration(string configurationName)
        {
            SampleProfileUdpBindingCollectionElement section = (SampleProfileUdpBindingCollectionElement)ConfigurationManager.GetSection(UdpConstants.UdpBindingSectionName);
            SampleProfileUdpBindingConfigurationElement element = section.Bindings[configurationName];
            if (element == null)
            {
                throw new ConfigurationErrorsException(string.Format(System.Globalization.CultureInfo.CurrentCulture,
                    "There is no binding named {0} at {1}.", configurationName, section.BindingName));
            }
            else
            {
                element.ApplyConfiguration(this);
            }
        }

        void Initialize()
        {
            transport = new UdpTransportBindingElement();
            session = new ReliableSessionBindingElement();
            compositeDuplex = new CompositeDuplexBindingElement();
            encoding = new TextMessageEncodingBindingElement();
        }

        //initialize a SampleProfileUdpBinding from the info collected in a ReliableSessionBindingElement if one is present.
        void InitializeFrom(UdpTransportBindingElement udpTransportBindingElement,
                    TextMessageEncodingBindingElement textMessageEncodingBindingElement,
                    ReliableSessionBindingElement reliableSessionBindingElement,
                    CompositeDuplexBindingElement compositeDuplexBindingElement)
        {
            this.transport.Multicast = udpTransportBindingElement.Multicast;
            this.transport.MaxBufferPoolSize = udpTransportBindingElement.MaxBufferPoolSize;
            this.transport.MaxReceivedMessageSize = udpTransportBindingElement.MaxReceivedMessageSize;

            ((TextMessageEncodingBindingElement)this.encoding).WriteEncoding = textMessageEncodingBindingElement.WriteEncoding;
            textMessageEncodingBindingElement.ReaderQuotas.CopyTo(((TextMessageEncodingBindingElement)this.encoding).ReaderQuotas);

            this.ReliableSessionEnabled = reliableSessionBindingElement != null;

            if (reliableSessionBindingElement != null)
            {
                this.SessionInactivityTimeout = reliableSessionBindingElement.InactivityTimeout;
                this.OrderedSession = reliableSessionBindingElement.Ordered;
            }

            if (compositeDuplexBindingElement != null)
            {
                this.ClientBaseAddress = compositeDuplexBindingElement.ClientBaseAddress;
            }
        }

        //try to create a SampleProfileUdpBinding from the collection of BindingElement
        //returns true if it is possible, with the resulting binding.
        public static bool TryCreate(BindingElementCollection elements, out Binding binding)
        {
            binding = null;
            if (elements.Count > 4)
            {
                return false;
            }

            ReliableSessionBindingElement reliableSessionBindingElement = null;
            CompositeDuplexBindingElement compositeDuplexBindingElement = null;
            TextMessageEncodingBindingElement textMessageEncodingBindingElement = null;
            UdpTransportBindingElement udpTransportBindingElement = null;

            foreach (BindingElement element in elements)
            {
                if (element is CompositeDuplexBindingElement)
                {
                    compositeDuplexBindingElement = element as CompositeDuplexBindingElement;
                }
                else if (element is TransportBindingElement)
                {
                    udpTransportBindingElement = element as UdpTransportBindingElement;
                }
                else if (element is TextMessageEncodingBindingElement)
                {
                    textMessageEncodingBindingElement = element as TextMessageEncodingBindingElement;
                }
                else if (element is ReliableSessionBindingElement)
                {
                    reliableSessionBindingElement = element as ReliableSessionBindingElement;
                }
                else
                {
                    return false;
                }
            }

            if (udpTransportBindingElement == null)
            {
                return false;
            }
            if (textMessageEncodingBindingElement == null)
            {
                return false;
            }

            if (((reliableSessionBindingElement != null) && (compositeDuplexBindingElement == null))
                || ((reliableSessionBindingElement == null) && (compositeDuplexBindingElement != null)))
            {
                return false;
            }

            SampleProfileUdpBinding sampleProfileUdpBinding = new SampleProfileUdpBinding();
            sampleProfileUdpBinding.InitializeFrom(udpTransportBindingElement, textMessageEncodingBindingElement,
                                            reliableSessionBindingElement, compositeDuplexBindingElement);
            if (!sampleProfileUdpBinding.IsBindingElementsMatch(udpTransportBindingElement, textMessageEncodingBindingElement,
                                                         reliableSessionBindingElement, compositeDuplexBindingElement))
            {
                return false;
            }

            binding = sampleProfileUdpBinding;
            return true;
        }

        bool IsBindingElementsMatch(UdpTransportBindingElement udpTransportBindingElement,
                    TextMessageEncodingBindingElement textMessageEncodingBindingElement,
                    ReliableSessionBindingElement reliableSessionBindingElement,
                    CompositeDuplexBindingElement compositeDuplexBindingElement)
        {
            if (!IsTransportMatch(this.transport, udpTransportBindingElement))
            {
                return false;
            }

            if(!IsEncodingMatch(this.encoding, textMessageEncodingBindingElement))
            {
                return false;
            }

            if (this.ReliableSessionEnabled)
            {
                if (!IsSessionMatch(this.session, reliableSessionBindingElement))
                {
                    return false;
                }
                if (compositeDuplexBindingElement != null)
                {
                    if (!IsCompositeDuplexMatch(this.compositeDuplex, compositeDuplexBindingElement))
                    {
                        return false;
                    }
                }
                else
                {
                    return false;
                }
            }
            else if (reliableSessionBindingElement != null)
            {
                return false;
            }

            return true;
        }

        bool IsTransportMatch(BindingElement a, BindingElement b)
        {
            if (b == null)
            {
                return false;
            }

            UdpTransportBindingElement transportA = a as UdpTransportBindingElement;
            UdpTransportBindingElement transportB = b as UdpTransportBindingElement;

            if (transportB == null)
            {
                return false;
            }
            if (transportA.MaxBufferPoolSize != transportB.MaxBufferPoolSize)
            {
                return false;
            }
            if (transportA.MaxReceivedMessageSize != transportB.MaxReceivedMessageSize)
            {
                return false;
            }
            if (transportA.Multicast != transportB.Multicast)
            {
                return false;
            }

            return true;
        }

        bool IsEncodingMatch(BindingElement a, BindingElement b)
        {
            if (b == null)
            {
                return false;
            }

            MessageEncodingBindingElement messageEncodingBindingElement = b as MessageEncodingBindingElement;
            if (messageEncodingBindingElement == null)
            {
                return false;
            }

            TextMessageEncodingBindingElement textA = a as TextMessageEncodingBindingElement;
            TextMessageEncodingBindingElement textB = b as TextMessageEncodingBindingElement;
            if (textB == null)
            {
                return false;
            }
            if (textA.MaxReadPoolSize != textB.MaxReadPoolSize)
            {
                return false;
            }
            if (textA.MaxWritePoolSize != textB.MaxWritePoolSize)
            {
                return false;
            }

            // compare XmlDictionaryReaderQuotas
            if (textA.ReaderQuotas.MaxStringContentLength != textB.ReaderQuotas.MaxStringContentLength)
            {
                return false;
            }
            if (textA.ReaderQuotas.MaxArrayLength != textB.ReaderQuotas.MaxArrayLength)
            {
                return false;
            }
            if (textA.ReaderQuotas.MaxBytesPerRead != textB.ReaderQuotas.MaxBytesPerRead)
            {
                return false;
            }
            if (textA.ReaderQuotas.MaxDepth != textB.ReaderQuotas.MaxDepth)
            {
                return false;
            }
            if (textA.ReaderQuotas.MaxNameTableCharCount != textB.ReaderQuotas.MaxNameTableCharCount)
            {
                return false;
            }

            if (textA.WriteEncoding.EncodingName != textB.WriteEncoding.EncodingName)
            {
                return false;
            }
            if(!IsMessageVersionMatch(textA.MessageVersion, textB.MessageVersion))
            {
                return false;
            }

            return true;
        }

        bool IsMessageVersionMatch(MessageVersion a, MessageVersion b)
        {
            if (b == null)
            {
                throw new ArgumentNullException("b");
            }
            if (a.Addressing == null)
            {
                throw new InvalidOperationException("MessageVersion.Addressing cannot be null");
            }

            if (a.Envelope != b.Envelope)
            {
                return false;
            }
            //if (a.Addressing.Namespace != b.Addressing.Namespace)
            //{
            //    return false;
            //}

            return true;
        }

        bool IsSessionMatch(BindingElement a, BindingElement b)
        {
            if (b == null)
            {
                return false;
            }

            ReliableSessionBindingElement sessionA = a as ReliableSessionBindingElement;
            ReliableSessionBindingElement sessionB = b as ReliableSessionBindingElement;

            if (sessionB == null)
            {
                return false;
            }
            if (sessionA.AcknowledgementInterval != sessionB.AcknowledgementInterval)
            {
                return false;
            }
            if (sessionA.FlowControlEnabled != sessionB.FlowControlEnabled)
            {
                return false;
            }
            if (sessionA.InactivityTimeout != sessionB.InactivityTimeout)
            {
                return false;
            }
            if (sessionA.MaxPendingChannels != sessionB.MaxPendingChannels)
            {
                return false;
            }
            if (sessionA.MaxRetryCount != sessionB.MaxRetryCount)
            {
                return false;
            }
            if (sessionA.MaxTransferWindowSize != sessionB.MaxTransferWindowSize)
            {
                return false;
            }
            if (sessionA.Ordered != sessionB.Ordered)
            {
                return false;
            }

            return true;
        }

        bool IsCompositeDuplexMatch(BindingElement a, BindingElement b)
        {
            if (b == null)
            {
                return false;
            }

            CompositeDuplexBindingElement duplexA = a as CompositeDuplexBindingElement;
            CompositeDuplexBindingElement duplexB = b as CompositeDuplexBindingElement;
            if (duplexB == null)
            {
                return false;
            }

            return (duplexB.ClientBaseAddress == duplexA.ClientBaseAddress);
        }
    }
}
